### PR TITLE
feat: Monitoring kafka integration

### PR DESCRIPTION
**Pull request by Vaishnavi190900**
Added metrics related to kafka messages success an failures using statsd exporter and prometheus.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -20,6 +20,10 @@ type GRPCConfig struct {
 	MaxSendMsgSizeInMB int `default:"10"`
 }
 
+type ProducerConfig struct {
```
Comment (rajuGT) [inline config/config.go]: MetricProducerConfig ?

File: config/config.yaml [hunk 2]
```diff
@@ -23,3 +23,6 @@ newrelic:
 db:
   # Connection string for postgres database
   connectionstring: "postgres://postgres@localhost:5432/db"
+
+producer:
```
Comment (rajuGT) [inline config/config.yaml]: metric_producer?

File: kafkaIntegration/producer.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+type KafkaProducer struct {
+	hostName string
+	producer *kafka.Producer
+	metrics  *ProducerMetrics
+}
+
+func NewKafkaProducer(hostName string) (*KafkaProducer, error) {
+	producer, err := kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": hostName})
+	if err != nil {
+		log.Printf("failed to create producer: %w", err)
```
Comment (rajuGT) [inline kafkaIntegration/producer.go]: to address the github mentioned error below, use `%s` and param err.Error()

File: kafkaIntegration/producer.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+	return &KafkaProducer{producer: producer, hostName: hostName, metrics: &ProducerMetrics{
+		SuccessCount: 0,
+		FailureCount: 0,
+		Latency:      0,
+	}}, nil
+}
+
+// ProduceMessages sends a message to the specified Kafka topic
+func (kp *KafkaProducer) ProduceMessages(topic string, protoMessage proto.Message, statsdClient statsd.Statter) error {
```
Comment (rajuGT) [inline kafkaIntegration/producer.go]: Name this method as ProduceMessageWithRetries and accept the retry and retryInterval as parameters.

Move code from line44 to line81 and make a method `func (kp *kafkaProducer) ProduceMessages.

Comment (rajuGT) [inline kafkaIntegration/producer.go]: Should we call Produce or Push?

File: kafkaIntegration/producer.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	for i := 1; i <= maxRetries; i++ {
+		message := &kafka.Message{
+			TopicPartition: kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny},
+			Value:          messageBytes,
+		}
+
+		deliveryChan := make(chan kafka.Event)
+		err = kp.producer.Produce(message, deliveryChan)
+		if err != nil {
+			fmt.Printf("failed to produce message (attempt %d/%d): %v\n", i, maxRetries, err)
```
Comment (rajuGT) [inline kafkaIntegration/producer.go]: remove fmt, it should be log

File: kafkaIntegration/producer.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+			time.Sleep(retryInterval)
+			continue
+		}
+
+		deliveryReport := <-deliveryChan
+		m := deliveryReport.(*kafka.Message)
+
+		if m.TopicPartition.Error != nil {
+			fmt.Printf("delivery failed (attempt %d/%d): %v\n", i, maxRetries, m.TopicPartition.Error)
```
Comment (rajuGT) [inline kafkaIntegration/producer.go]: remove fmt, it should be log

File: kafkaIntegration/producer.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+		endTime := time.Now()
+		latency := endTime.Sub(startTime)
+		kp.metrics.Latency = latency
+
+		kp.metrics.SuccessCount++
+
+		err := statsdClient.Gauge("kafka_integration.latency", latency.Milliseconds(), 1)
+		if err != nil {
+			return err
```
Comment (rajuGT) [inline kafkaIntegration/producer.go]: Failed to send metrics shouldn't stop the business flow, you can log this error and ignore, i.e. no need to return err.

File: kafkaIntegration/producer.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+		}
+
+		err = statsdClient.Inc("kafka_integration.success_count", int64(kp.metrics.SuccessCount), 1)
+		if err != nil {
+			return err
```
Comment (rajuGT) [inline kafkaIntegration/producer.go]: Failed to send metrics shouldn't stop the business flow, you can log this error and ignore, i.e. no need to return err.

File: kafkaIntegration/producer.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+			continue
+		}
+
+		endTime := time.Now()
+		latency := endTime.Sub(startTime)
+		kp.metrics.Latency = latency
+
+		kp.metrics.SuccessCount++
+
+		err := statsdClient.Gauge("kafka_integration.latency", latency.Milliseconds(), 1)
```
Comment (rajuGT) [inline kafkaIntegration/producer.go]: define these constants as const "kafka_integration.latency"

**Review by rajuGT — CHANGES_REQUESTED** [review]

File: config/config.go (hunk 1 above)
Comment (rajuGT) [reply config/config.go]: rename it as KafkaProducerConfig

File: config/config.go [hunk 10]
```diff
@@ -29,4 +33,5 @@ type Config struct {
 	GRPC          GRPCConfig
 	NewRelic      NewRelicConfig
 	DB            DBConfig
+	Producer      ProducerConfig
```
Comment (rajuGT) [inline config/config.go]: KafkaProducer

File: config/config.yaml (hunk 2 above)
Comment (rajuGT) [reply config/config.yaml]: kafka_producer

File: kafkaIntegration/monitoring.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
+package kafkaIntegration
```
Comment (rajuGT) [inline kafkaIntegration/monitoring.go]: make a package changeEventProducer and inside that lets have package kafka.

File: kafkaIntegration/producer.go [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+	}}, nil
+}
+
+func (kp *KafkaProducer) PushMessagesWithRetries(topic string, protoMessage proto.Message, statsdClient statsd.Statter, retries int, retryInterval time.Duration) error {
+	messageBytes, err := proto.Marshal(protoMessage)
+	if err != nil {
+		return fmt.Errorf("failed to marshal Protobuf message: %v", err)
+	}
+	startTime := time.Now()
+	for i := 1; i <= retries; i++ {
```
Comment (rajuGT) [inline kafkaIntegration/producer.go]: prefer starting with zero

File: changeEventProducer/kafka/producer.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+		return err
+	}
+
+	deliveryReport := <-deliveryChan
+	m := deliveryReport.(*kafka.Message)
+
+	if m.TopicPartition.Error != nil {
+		log.Printf(err.Error())
+		return err
+	}
```
Comment (manigandand) [inline changeEventProducer/kafka/producer.go]: plz validate the typecast with ok flag to make sure we don't panic upon unsupported types.

```suggestion
	if m, ok := deliveryReport.(*kafka.Message); ok && m.TopicPartition.Error != nil {
		log.Printf(err.Error())
		return err
	}
```

**Review by manigandand — CHANGES_REQUESTED** [review]
please make these changes

**Review by manigandand — APPROVED** [review]
Please resolve the lint issue and merge conflicts @Vaishnavi190900

**Review by manigandand — CHANGES_REQUESTED** [review]
Please resolve the lint issue and merge conflicts @Vaishnavi190900

**Comment by Vaishnavi190900** [comment]
@rajuGT @manigandand resolved all the comments and checks are also passed. Can you please review and approve?

**Comment by manigandand** [comment]
LGTM

**Review by manigandand — APPROVED** [review]

File: changeEventProducer/kafka/producer.go [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+func (kp *KafkaProducer) PushMessagesWithRetries(topic string, protoMessage proto.Message, retries int, retryInterval time.Duration) error {
+	messageBytes, err := proto.Marshal(protoMessage)
+	if err != nil {
+		return fmt.Errorf("failed to marshal Protobuf message: %v", err)
+	}
+
+	for i := 0; i < retries; i++ {
+		err := kp.PushMessages(messageBytes, topic)
+		if err != nil {
+			time.Sleep(retryInterval)
```
Comment (rajuGT) [inline changeEventProducer/kafka/producer.go:47]: can we increase the failurecount metric here?

File: changeEventProducer/kafka/producer.go [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+		return nil, err
+	}
+
+	return &KafkaProducer{producer: producer, hostName: hostName, statsdClient: statsdClient}, nil
+}
+
+func (kp *KafkaProducer) PushMessagesWithRetries(topic string, protoMessage proto.Message, retries int, retryInterval time.Duration) error {
+	messageBytes, err := proto.Marshal(protoMessage)
+	if err != nil {
+		return fmt.Errorf("failed to marshal Protobuf message: %v", err)
```
Comment (rajuGT) [inline changeEventProducer/kafka/producer.go]: %s, err.Error()

**Review by rajuGT — CHANGES_REQUESTED** [review]

**Review by rajuGT — APPROVED** [review]

**Pull request merged**